Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,886 @@
+      // element, return null.
+      while (ancestors.Count != 0) {
+        var maxAmongAncestors = ancestors.Max(Height);
+        var highestAncestors = ancestors.Where(ancestor => Height(ancestor) == maxAmongAncestors).ToList();
+        if (highestAncestors.Count == 1) {
+          return highestAncestors.ElementAt(0);
+        }
+        ancestors.RemoveAll(ancestor => Height(ancestor) == maxAmongAncestors);
+      }
+      return null;
+    }
+
+    TopLevelDecl/*?*/ MeetHeads(TopLevelDecl a, TopLevelDecl b) {
+      Contract.Requires(a != null);
+      Contract.Requires(b != null);
+      var aAncestors = new HashSet<TopLevelDecl>();
+      if (aAncestors.Contains(b)) {
+        // that's good enough; let's pick a
+        return a;
+      }
+      var bAncestors = new HashSet<TopLevelDecl>();
+      if (bAncestors.Contains(a)) {
+        // that's good enough; let's pick b
+        return b;
+      }
+      // give up
+      return null;
+    }
+
+    IEnumerable<DPreType> AllSubBounds(PreTypeProxy proxy, ISet<PreTypeProxy> visited) {
+      Contract.Requires(proxy.PT == null);
+      if (!visited.Contains(proxy)) {
+        visited.Add(proxy);
+        foreach (var constraint in unnormalizedSubtypeConstraints) {
+          if (constraint.Super.Normalize() == proxy) {
+            var sub = constraint.Sub.Normalize();
+            if (sub is PreTypeProxy subProxy) {
+              foreach (var pt in AllSubBounds(subProxy, visited)) {
+                yield return pt;
+              }
+            } else {
+              yield return (DPreType)sub;
+            }
+          }
+        }
+      }
+    }
+
+    IEnumerable<DPreType> AllSuperBounds(PreTypeProxy proxy, ISet<PreTypeProxy> visited) {
+      Contract.Requires(proxy.PT == null);
+      if (!visited.Contains(proxy)) {
+        visited.Add(proxy);
+        foreach (var constraint in unnormalizedSubtypeConstraints) {
+          if (constraint.Sub.Normalize() == proxy) {
+            var super = constraint.Super.Normalize();
+            if (super is PreTypeProxy superProxy) {
+              foreach (var pt in AllSuperBounds(superProxy, visited)) {
+                yield return pt;
+              }
+            } else {
+              yield return (DPreType)super;
+            }
+          }
+        }
+      }
+    }
+
+
+    IEnumerable<PreType> DirectionalBounds(PreTypeProxy tproxy, int direction) {
+      foreach (var constraint in unnormalizedSubtypeConstraints) {
+        if (0 <= direction && constraint.Super.Normalize() == tproxy) {
+          yield return constraint.Sub;
+        }
+        if (direction <= 0 && constraint.Sub.Normalize() == tproxy) {
+          yield return constraint.Super;
+        }
+      }
+    }
+
+    // ---------------------------------------- Guarded constraints ----------------------------------------
+
+    private List<Func<bool>> guardedConstraints = new();
```

Comment:
The division between equality, subtype and guarded constraints, the last of which seems to contain many types of constraints, seems somewhat arbitrary to me.

The only one of the three that stands out to me are subtype constraints since they're also used as an information store.

Reply:
I tried to reduce the number of such categories as much as possible, and the current division is what I came up with.

With a priority queue (along your suggestion in another comment above), it may be possible to combine these constraints further.

The way I've been thinking of a "guarded constraint" is that "if the guard evaluates the true, then the 'guarded constraint' turns into an actual constraint". Perhaps a name like "constraint that's may or may not materialize" would be more to the point.

The last category, "post-inference confirmations", is one of the conceptual clarifications from the previous type-inference system in Dafny. Previously, there was an attempt to try to make every piece of information help the inference along, even though some information never did. As a (possibly fictitious) illustration, there could be a constraint "bitvector or integer" and a constraint "bitvector or char", and then one could, in principle, take the intersection of those constraints and figure out that something is a bitvector. In the new type-inference system, those kinds of constraints are made into "post-inference confirmations", which means they don't aid the _inference_ at all, but are just there to be checked after inference has done its thing.